Unify closure serializing logic for `to nuon`, `to msgpack`, and `to json`

132ikl opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->

Before this PR, `to msgpack`/`to msgpackz` and `to json` serialize closures as `nil`/`null` respectively, when the `--serialize` option isn't passed. This PR makes it an error to serialize closures to msgpack or JSON without the `--serialize` flag, which is the behavior of `to nuon`.

This PR also adds the `--serialize` flag to `to msgpack`.

This PR also changes `to nuon` and `to json` to return an error if they cannot find the block contents of a closure, rather than serializing an empty string or an error string, respectively. This behavior is replicated for `to msgpack`.

 It also changes `to nuon`'s error message for serializing closures without `--serialize` to be the same as the new errors for `to json` and `to msgpack`.

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->
* Add `--serialize` flag to `to msgpack`, similar to the `--serialize` flag for `to nuon` and `to json`.
* Serializing closures to JSON or msgpack without `--serialize`

Partially fixes #11738

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->
N/A

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->
N/A

fdncred commented:
`to nuon` already has the `--serialize` flag. Oh, I bet you mean `to msgpack`. I definitely didn't add it there.

fdncred reviewed (approved):
LGTM

sholderbach commented on crates/nu-command/src/formats/to/msgpack.rs:
```diff
@@ -208,25 +228,53 @@ pub(crate) fn write_value(
-                .err_span(span)?;
+        Value::Closure { val, .. } => {
+            if serialize_types {
+                let block = engine_state.get_block(val.block_id);
+                let closure_string: &str = if let Some(span) = block.span {
+                    let contents_bytes = engine_state.get_span_contents(span);
+                    &String::from_utf8_lossy(contents_bytes)
+                } else {
+                    return Err(WriteError::Shell(Box::new(ShellError::CantConvert {
+                        to_type: "string".into(),
+                        from_type: "closure".into(),
+                        span,
+                        help: Some(format!(
+                            "unable to retrieve block contents for closure with id {}",
+                            val.block_id.get()
+                        )),
+                    })));
+                };
```
Could this part you repeated across a few places live on the closure type?

e.g. `Closure::coerce_to_string`

132ikl replied:
that sounds good to me, I thought of doing something like that but wasn't sure where to put it. not sure if `Closure` is really the _best_ place to put it, but it keeps it out of the other stringifying logic so it's good enough for me. we can probably revisit this as part of #15286 and #14090

sholderbach replied:
I like on the type more than in a large `match` like on `Value`. Could also be useful if you want to debug print stuff in places like the engine.

With engine_state should be fine in `nu-protocol`

sholderbach reviewed (commented):
That description has a lot of "also"s for stuff not subsumed by the title.

132ikl commented on crates/nu-command/src/formats/to/json.rs:
```diff
@@ -78,16 +78,12 @@ impl Command for ToJson {
-                    span,
-                    help: None,
-                },
+            _ => Err(ShellError::CantConvert {
```
> The problem is preexiting

could you elaborate on this, I'm not quite following

> but this use of CantConvert irks me

honestly same, I'm willing to change them to something else. do you have any suggestions?

sholderbach replied:
That is probably more a young man yells at cloud moment, I don't remember us having a specific variant for file format / spec intricacies.

It was like this before, so I don't want to force you to make a change in this PR, we can look at this when chipping away at the whole TypeError complex.

132ikl commented:
draft mode to address stefan's comments

sholderbach commented:
Do we want to ship this with `0.103.0`? 
Addressing my request for a helper should be relatively straightforward. Or are there any other concerns after our last meeting?

132ikl commented:
> Do we want to ship this with 0.103.0

all good with me. just moved the string coercion code into `impl Closure`.

also, I renamed `capture_block.rs` to `closure.rs` since the only thing in there now is `Closure` itself, and `CaptureBlock` was renamed to `Closure` a long time ago

sholderbach reviewed (approved):
Nice

fdncred commented:
I'm wondering now if `to text` should have the same treatment.

132ikl commented:
> I'm wondering now if to text, to toml, to yaml should have the same treatment.

yes, definitely. also kind of along the same lines as #15286
